### PR TITLE
librealsense: refactor to allow use of cuda, and prepare for python integration

### PR DESCRIPTION
**Pull request by freezeboy**
###### Motivation for this change

librealsense can be compiled to use CUDA to improve some algorithms.
This patch adds two conditional parametes to add python support and cuda
support

also adds two variations (like tensorflow does):

* pkgs.librealsense (original, uses config.cudaSupport dynamically)
* pkgs.librealsenseWithCuda - forces cudaSupport
* pkgs.librealsenseWithoutCuda - forces no cudaSupport

and for python:

* pkgs.python3Packages.pyrealsense2
* pkgs.python3Packages.pyrealsense2WithCuda
* pkgs.python3Packages.pyrealsense2WithoutCuda

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by freezeboy** [comment]
Cuda versions are currently subject to this bug: https://github.com/IntelRealSense/librealsense/issues/6573#issuecomment-643725732
I don't plan to apply the patch as another PR is proposing an update to 2.36

**Review by brian-dawn — APPROVED** [review]
I don't have an easy way to test these changes but it looks good to me!

**Comment by brian-dawn** [comment]
@freezeboy thanks for the changes. Since I don't use cuda with librealsense would you be interested in also helping maintain the librealsense package?

**Comment by freezeboy** [comment]
Yeah, no problem on my side even if I don't use it a lot either I can help you

File: pkgs/top-level/all-packages.nix
Comment (jonringer) [inline pkgs/top-level/all-packages.nix:17505]: I believe the current convention is just to have no suffix
```suggestion
  librealsense = callPackage ../development/libraries/librealsense {
```

File: pkgs/development/libraries/librealsense/default.nix
Comment (jonringer) [inline pkgs/development/libraries/librealsense/default.nix:3]: looks like you don't care so much for the packageset, as much as python
```suggestion
, enablePython ? false, python }:
```

File: pkgs/development/libraries/librealsense/default.nix
Comment (jonringer) [inline pkgs/development/libraries/librealsense/default.nix:24]:
```suggestion
    ++ lib.optional enablePython python;
```

File: pkgs/development/libraries/librealsense/default.nix
Comment (jonringer) [inline pkgs/development/libraries/librealsense/default.nix:37]:
```suggestion
    mkdir -p $out/${python.sitePackages}/pyrealsense2
```

File: pkgs/development/libraries/librealsense/default.nix
Comment (jonringer) [inline pkgs/development/libraries/librealsense/default.nix:38]:
```suggestion
    mv $out/lib/*cpython*.so* $out/${python.sitePackages}
```

File: pkgs/top-level/python-packages.nix
Comment (jonringer) [inline pkgs/top-level/python-packages.nix:2533]: `python` gets aliased to the correct version within a python package set
```suggestion
```

File: pkgs/top-level/python-packages.nix
Comment (jonringer) [inline pkgs/top-level/python-packages.nix:2538]:
```suggestion
```

File: pkgs/top-level/python-packages.nix
Comment (jonringer) [inline pkgs/top-level/python-packages.nix:2541]:
```suggestion
  pyrealsense2 = toPythonModule (pkgs.librealsenseWithoutCuda.override {
```

File: pkgs/top-level/python-packages.nix
Comment (jonringer) [inline pkgs/top-level/python-packages.nix:2543]:
```suggestion
```

File: pkgs/development/libraries/librealsense/default.nix
Comment (freezeboy) [reply pkgs/development/libraries/librealsense/default.nix:3]: in fact it should compile a python module, so it should basically call buildPythonPackage, except that the job is done by their cmake process instead of a setup.py file

File: pkgs/top-level/all-packages.nix
Comment (freezeboy) [reply pkgs/top-level/all-packages.nix:17505]: the "default" version uses config to switch, whereas this one forces to always "not" use cuda

**Review by brian-dawn — APPROVED** [review]

**Comment by tpwrules** [comment]
I made my own pyrealsense2 PR (mentioned above) without knowing this one existed.

Unfortunately, Intel changed the way they install the Python bindings with 2.36.0. Now that nixpkgs has moved to that version, this PR no longer builds using `nixpkgs-review -pr 92997`. You are welcome to take the patches I made in my PR to make yours work with 2.36.0. I don't have a way to test with CUDA, so it doesn't make sense to do it the other way.

I have a further patch to enable T265 support that will require consultation with the maintainers due to licensing issues. Not sure the best way to do that, but regardless I was going to wait until Python support was merged.

**Comment by freezeboy** [comment]
I'll look at your method as I didn't check this update problem

**Comment by freezeboy** [comment]
About the examples, I don't particularly care about building them as I don't need them, aren't they big ? should I add a flag to let the user remove them or it is not worth it ?

**Comment by freezeboy** [comment]
rs-color, rs-depth and rs-distance cannot find some symbols from stdcpp library, e.g. :
librealsense2.so.2.36: undefined symbol: _ZNSt7__cxx1119basic_ostringstreamIcSt11char_traitsIcESaIcEEC1Ev

Do you know how this package is named in nix ?

**Comment by tpwrules** [comment]
Following up on our conversation on IRC:

1. You said you got the undefined symbol error resolved?

2. The compiled command line example binaries sum to under 2 megabytes and are useful for testing and troubleshooting. I think they should be unconditionally built as there's no real penalty.

3. Regarding the required Python packages, I believe you are correct, only python is required. Both setuptools and wheel can be removed. They were remnants of earlier experiments.

**Comment by tpwrules** [comment]
I reviewed the latest commit with nixpkgs-review and the examples build and run fine. The Python module is built fine, but it doesn't appear to Python. I believe this is because `python` should be moved to `propagatedBuildInputs` so `toPythonModule` picks the module up.

After doing `nix-shell -p nixpkgs-review --run "nixpkgs-review pr 92997 -p python38Packages.pyrealsense2"`, you should be able to print out any attached RealSense devices using the examples with `results/python38Packages.pyrealsense2/bin/rs-enumerate-devices` or with Python using `python3 -c 'import pyrealsense2 as rs; print(list(rs.context().query_devices()))'` (if there are no attached devices but it otherwise worked correctly, it will just print an empty list `[]`).

**Comment by freezeboy** [comment]
what do you mean it doesn't appear to python, it looks like it is ok on my side ..

**Comment by tpwrules** [comment]
We discussed on IRC and there was a misunderstanding on my end.

I'm happy with this PR and think it should be merged. I'll close mine out.

File: pkgs/development/libraries/librealsense/default.nix
Comment (brian-dawn) [inline pkgs/development/libraries/librealsense/default.nix:27]: Are wheel and setuptools still needed?

**Review by brian-dawn — APPROVED** [review]

File: pkgs/development/libraries/librealsense/default.nix
Comment (freezeboy) [reply pkgs/development/libraries/librealsense/default.nix]: True, they are no more useful, i removed them

**Review by brian-dawn — APPROVED** [review]

**Review by brian-dawn — APPROVED** [review]

**Review by jonringer — APPROVED** [review]
diff LGTM, but only did a quick glance:

```
https://github.com/NixOS/nixpkgs/pull/92997
8 packages built:
librealsense librealsenseWithCuda python27Packages.pyrealsense2 python27Packages.pyrealsense2WithCuda python37Packages.pyrealsense2 python37Packages.pyrealsense2WithCuda python38Packages.pyrealsense2 python38Packages.pyrealsense2WithCuda
```

**Comment by jonringer** [comment]
ping me tomorrow if no one merges

**Comment by freezeboy** [comment]
well ping @jonringer

**Review by jonringer — APPROVED** [review]
```
https://github.com/NixOS/nixpkgs/pull/92997
8 packages built:
librealsense librealsenseWithCuda python27Packages.pyrealsense2 python27Packages.pyrealsense2WithCuda python37Packages.pyrealsense2 python37Packages.pyrealsense2WithCuda python38Packages.pyrealsense2 python38Packages.pyrealsense2WithCuda
```

**Pull request merged**
